java-checks/src/main/java/org/sonar/java/checks/tests/MockitoArgumentMatchersUsedOnAllParametersCheck.java

```diff
@@ -104,4 +130,30 @@ private static ExpressionTree skipCasts(ExpressionTree tree) {
     }
     return current;
   }
+
+  private static class MethodVisitor extends BaseTreeVisitor {
+    static Map<MethodTree, Boolean> cachedResults = new HashMap<>();
+    boolean returnsAnArgumentMatcher = false;
+
+    @Override
+    public void visitMethod(MethodTree tree) {
+      if (cachedResults.containsKey(tree)) {
+        returnsAnArgumentMatcher = cachedResults.get(tree);
+        return;
+      }
+      cachedResults.put(tree, Boolean.FALSE);
+      List<ExpressionTree> terminalMethodInvocations = tree.block().body().stream()
+        .filter(statement -> statement.is(Tree.Kind.RETURN_STATEMENT))
+        .map(ReturnStatementTree.class::cast)
+        .map(ReturnStatementTree::expression)
+        .filter(expression -> expression.is(Tree.Kind.METHOD_INVOCATION))
+        .collect(Collectors.toList());
+      if (terminalMethodInvocations.isEmpty()) {
+        return;
+      }
+      returnsAnArgumentMatcher = terminalMethodInvocations.stream()
+        .allMatch(MockitoArgumentMatchersUsedOnAllParametersCheck::isArgumentMatcherLike);
```

Comment:
As discussed, recursively searching for an argument matcher is a bit overkill here. An approximation that seems good enough would be to consider any method call at this point as an argument matcher.

This will also simplify the implementation.